keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1055 @@
+This is the case with `client-go` today. In this scenario, if a request times out, 
+the client is most likely to receive a `context deadline exceeded` error due to
+the fact that the client will see the `Context` of the client request be expired
+well before the `504 GatewayTimeout` response from the server makes the round-trip
+to the client.
+
+Also, the client can not control when or whether the request handler writes to
+the `ResponseWriter` object. So, today a client can not deterministically catch
+the `504 GatewayTimeout` error from the server when a request times out.
+
+With the enablement of this feature, and the removal of the `timeout` filter the
+apiserver is no longer sending a `504 GatewayTimeout` response when a request 
+times out.
+
+The following goes through different scenarios, and for each we describe how the
+server and client behaves differently when this feature is enabled.
+
+
+Scenario `A`:
+- client specifies a timeout in the `timeout` parameter: **No**
+- request handler writes to the `ResponseWriter` object before timeout happens: **No**
+
+*Observation*:
+
+`PerHandlerReadWriteTimeout` - **Disabled**:
+- client: receives an `http.StatusGatewayTimeout` status code
+- client: reading the `Body` of the `Response` object does not yield any error
+- server: write to the `ResponseWriter` yields an `http: Handler timeout` errorg immediately
+
+`PerHandlerReadWriteTimeout` - **Enabled**:
+- client: http/2.0 client receives a stream reset error, on the other hand, http/1x 
+  client receives a `local error: tls: bad record MAC` error.
+- server: write to the `ResponseWriter` yields an `i/o timeout` error once the
+  underlying buffer is full, so it may take a few `Write` invocations before the
+  handler sees the timeout error.
+
+> Same behavior for both http/1x and http/2.0
+
+<br>Scenario `B`:
+- client specifies a timeout in the `timeout` parameter: **No**
+- request handler writes to the ResponseWriter object before timeout happens: **Yes**
+
+*Observation* (with http/2.0): 
+
+`PerHandlerReadWriteTimeout` - **Disabled**:
+- client: receives an http2 stream reset error, since the client-facing 
+  goroutine in the `timeout` filter panics if the `ResponseWriter` object has
+  already been written to.
+- server: write to the `ResponseWriter` yields an `http: Handler timeout` error immediately
+
+
+`PerHandlerReadWriteTimeout` - **Enabled**
+- client: same as above (since net/http directly resets the http2 stream)
+- server: write to the `ResponseWriter` yields an `i/o timeout` error once the
+  underlying buffer is full.
+  
+
+*Observation* (with http/1x):
+
+`PerHandlerReadWriteTimeout` - **Disabled**:
+- client: receives an `EOF` error
+- server: write to the `ResponseWriter` yields an `http: Handler timeout` error immediately
+
+
+`PerHandlerReadWriteTimeout` - **Enabled**
+- client: receives a `local error: tls: bad record MAC` error
```

Comment:
that's a weird error.  I may need to re-look at these tomorrow.

Reply:
yes, I have a test that actually generates this error, I haven;t closely looked at it yet, I will open an issue with the golang team for this.